### PR TITLE
Fix #10624 - Fix Transition from 23.2.0 to 24.2.0 when a HeatExchanger:AirToAir:SensibleAndLatent has blank effectiveness fields

### LINKED ISSUE #10624: Transition to v24.1 crashes in the HX transition if there are blank effectiveness fields
Issue overview
--------------
Got a crash transitioning a user file from 23.2 to 24.1.
image

It's failing on this line in the source:
`                  READ(HxEffectAt75Airflow(i), *) effect75`

The incoming HX objects look like this:
```
HeatExchanger:AirToAir:SensibleAndLatent,
    VMC_1_HEX,               !- Name
    Heat_Recovery_Availability,  !- Availability Schedule Name
    0.305,                   !- Nominal Supply Air Flow Rate {m3/s}
    0.75,                    !- Sensible Effectiveness at 100% Heating Air Flow {dimensionless}
    ,                        !- Latent Effectiveness at 100% Heating Air Flow {dimensionless}
    0.75,                    !- Sensible Effectiveness at 75% Heating Air Flow {dimensionless}
    ,                        !- Latent Effectiveness at 75% Heating Air Flow {dimensionless}
    0.75,                    !- Sensible Effectiveness at 100% Cooling Air Flow {dimensionless}
    ,                        !- Latent Effectiveness at 100% Cooling Air Flow {dimensionless}
    0.75,                    !- Sensible Effectiveness at 75% Cooling Air Flow {dimensionless}
    ,                        !- Latent Effectiveness at 75% Cooling Air Flow {dimensionless}
```

Workaround
---------------
Fill in the default value of 0.0 in the blank HX effectiveness fields in the incoming v23.2 file.

### Details
Some additional details for this issue (if relevant):
 - Platform Win11
 - Version of EnergyPlus 24.1
 - Helpdesk ticket number 16730

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added in EnergyPlusDevSupport\DefectFiles
 - [x] Ticket added to EnergyPlus Defect Complexity (Github Project)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fix #10624
     - Adding a new function `GetFieldOrIDDDefault` and using that when retrieving HX effectiveness fields
     - This will effectively set the values to 0.0 if blank, since that's what the IDD default is for these fields.
 - Also avoid writting a Table:IndependentVariable + List when no Table:Lookup was written

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jmarrec** [comment]
Comparing 

| Original V2320 | Develop Transition | This PR Transition |
|----------------|--------------------|--------------------|


## FillHXBlanks_V2320.idf
 
**Notice the useless Table:IndependentVariable and Table:IndependentVariableList in develop but not this PR**

image



## NoBlanksDifferentEffectiveness_V2320.idf
image


## WithHXBlanks_V2320.idf

**develop segfaults**

image

## WithHXBlanks_Different_Effectiveness_V2320.idf

**develop segfaults**

image

File: src/Transition/InputProcessor.f90 [hunk 1]
```diff
@@ -4814,6 +4814,28 @@ FUNCTION IPTrimSigDigits(IntegerValue) RESULT(OutputString)
...
+  CHARACTER(len=LEN(InArgString)) :: ResultString       ! Result String
+
+  ResultString=InArgString
+  IF (ResultString == Blank) THEN
+    ResultString = FldDefaultString
+  ENDIF
+
+  RETURN
+
+END FUNCTION GetFieldOrIDDDefault
```
Comment (jmarrec) [inline src/Transition/InputProcessor.f90:4837]: New helper

File: src/Transition/CreateNewIDFUsingRulesV24_1_0.f90 [hunk 2]
```diff
@@ -429,14 +429,22 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
...
+                ! Latent Effectiveness at 75% Cooling Air Flow
+                HxEffectAt75Airflow(4) = GetFieldOrIDDDefault(InArgs(11), FldDefaults(11))
+                ! Sensible Effectiveness at 100% Heating Air Flow
+                HxEffectAt100Airflow(1) = GetFieldOrIDDDefault(InArgs(4), FldDefaults(4))
+                ! Latent Effectiveness at 100% Heating Air Flow
+                HxEffectAt100Airflow(2) = GetFieldOrIDDDefault(InArgs(5), FldDefaults(5))
+                ! Sensible Effectiveness at 100% Cooling Air Flow
+                HxEffectAt100Airflow(3) = GetFieldOrIDDDefault(InArgs(8), FldDefaults(8))
+                ! Latent Effectiveness at 100% Cooling Air Flow
+                HxEffectAt100Airflow(4) = GetFieldOrIDDDefault(InArgs(9), FldDefaults(9))
```
Comment (jmarrec) [inline src/Transition/CreateNewIDFUsingRulesV24_1_0.f90:447]: Use it.

File: src/Transition/CreateNewIDFUsingRulesV24_1_0.f90 [hunk 3]
```diff
@@ -138,7 +138,7 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
   CHARACTER(20), DIMENSION(4) :: HxEffectAt75Airflow
   CHARACTER(20), DIMENSION(4) :: HxEffectAt100Airflow
   CHARACTER(MaxNameLength + 2), DIMENSION(4) :: HxTableName
-  LOGICAL :: tableAdded
+  LOGICAL :: tableAdded = .false.
```
Comment (jmarrec) [inline src/Transition/CreateNewIDFUsingRulesV24_1_0.f90:141]: Avoid writting Table:IndependentVariable(List) when not needed

File: src/Transition/InputProcessor.f90 [hunk 4]
```diff
@@ -4814,6 +4814,28 @@ FUNCTION IPTrimSigDigits(IntegerValue) RESULT(OutputString)
 
 END FUNCTION IPTrimSigDigits
 
+FUNCTION GetFieldOrIDDDefault(InArgString, FldDefaultString)  RESULT (ResultString)
+
+          ! PURPOSE OF THIS SUBROUTINE:
+          ! If InArgString is Blank, replace with FldDefaultString
+
+  IMPLICIT NONE    ! Enforce explicit typing of all variables in this routine
```
Comment (Myoldmopar) [inline src/Transition/InputProcessor.f90:4822]: Implicit None is already declared at the top of the module, so probably not necessary, but totally fine.

File: src/Transition/CreateNewIDFUsingRulesV24_1_0.f90 [hunk 5]
```diff
@@ -429,14 +429,22 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
...
-                HxEffectAt75Airflow(1) = TRIM(InArgs(6))    ! Sensible Effectiveness at 75% Heating Air Flow
-                HxEffectAt75Airflow(2) = TRIM(InArgs(7))    ! Latent Effectiveness at 75% Heating Air Flow
-                HxEffectAt75Airflow(3) = TRIM(InArgs(10))   ! Sensible Effectiveness at 75% Cooling Air Flow
-                HxEffectAt75Airflow(4) = TRIM(InArgs(11))   ! Latent Effectiveness at 75% Cooling Air Flow
-                HxEffectAt100Airflow(1) = TRIM(InArgs(4))   ! Sensible Effectiveness at 100% Heating Air Flow
-                HxEffectAt100Airflow(2) = TRIM(InArgs(5))   ! Latent Effectiveness at 100% Heating Air Flow
-                HxEffectAt100Airflow(3) = TRIM(InArgs(8))   ! Sensible Effectiveness at 100% Cooling Air Flow
-                HxEffectAt100Airflow(4) = TRIM(InArgs(9))   ! Latent Effectiveness at 100% Cooling Air Flow
+                ! Sensible Effectiveness at 75% Heating Air Flow
+                HxEffectAt75Airflow(1) = GetFieldOrIDDDefault(InArgs(6), FldDefaults(6))
```
Comment (Myoldmopar) [inline src/Transition/CreateNewIDFUsingRulesV24_1_0.f90:433]: So FldDefaults will always be populated with meaningful or blank values?  I've never used that array so just confirming.

**Review by Myoldmopar — APPROVED** [review]
This seems cool, I've never paid attention to the FldDefaults before, but it looks look a great use of it.  And it's just one of the args passed to GetObjectDefInIDD, so it should be safe.  This is good stuff.

I'll do a quick test but this should merge soon.  @mjwitte am I missing anything here?

**Comment by mjwitte** [comment]
@jmarrec @Myoldmopar This works. Nothing to add.

**Comment by Myoldmopar** [comment]
Awesome, merging this.  Thanks @jmarrec

**Pull request merged**